File: python/clearings/src/clearings/values.py
from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass
from typing import Generic, TypeAlias, TypeVar, cast

from ._control import ErrorCode

Value: TypeAlias = None | bool | int | float | str | list["Value"] | dict[str, "Value"]
T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)
ACTIONS: dict[ErrorCode, str] = {
    "INVALID_PLAN": "Check flow references, dependency limits and policy settings.",
    "INVALID_VALUE": "Use supported portable values and the declared operation schema.",
    "UNSUPPORTED": "Use an implemented capability and matching SDK/native versions.",
    "CAPACITY": "Reduce admitted work or explicitly adjust the runtime capacity.",
    "OPERATION_FAILED": "Inspect the operation in the application; retry only under its declared semantics.",
    "TRANSFORM_FAILED": "Inspect the pure transformation at the indicated source.",
    "CANCELLED": "Start another invocation if the result is still needed.",
    "TIMEOUT": "Inspect queue and service time before changing the deadline.",
    "CLOSED": "Create a new Runtime for subsequent invocations.",
}


class ClearingsError(Exception):
    def __init__(self, code: ErrorCode, source: str | None = None) -> None:
        self.code = code
        self.source = source
        self.next_action = ACTIONS[code]
        super().__init__(f"{code}{' at ' + source if source else ''}. {self.next_action}")


def copy_value(input: T) -> T:
    """Snapshot portable data in the host; reject implicit codecs and coercion."""
    entries = 0
    byte_count = 0
    path: set[int] = set()

    def string(value: str) -> str:
        nonlocal byte_count
        try:
            byte_count += len(value.encode("utf-8"))
        except UnicodeEncodeError:
            raise ClearingsError("INVALID_VALUE") from None
        if byte_count > 1_048_576:
            raise ClearingsError("CAPACITY")
        return value

    def visit(value: object, depth: int) -> Value:
        nonlocal entries
        entries += 1
        if entries > 100_000 or depth > 64:
            raise ClearingsError("CAPACITY")
        if value is None or type(value) is bool:
            return cast(Value, value)
        if type(value) is str:
            return string(value)
        if type(value) is int:
            if abs(value) > 9_007_199_254_740_991:
                raise ClearingsError("INVALID_VALUE")
            return value
        if type(value) is float:
            number = value
            if not math.isfinite(number) or (
                number.is_integer() and abs(number) > 9_007_199_254_740_991
            ):
                raise ClearingsError("INVALID_VALUE")
            return 0 if number == 0 else number
        if type(value) not in (list, dict) or id(value) in path:
            raise ClearingsError("INVALID_VALUE")
        path.add(id(value))
        if type(value) is list:
            result: Value = [visit(item, depth + 1) for item in cast(list[object], value)]
        else:
            record: dict[str, Value] = {}
            for key, item in cast(dict[object, object], value).items():
                if type(key) is not str:
                    raise ClearingsError("INVALID_VALUE")
                record[string(key)] = visit(item, depth + 1)
            result = record
        path.remove(id(value))
        return result

    return cast(T, visit(input, 0))


@dataclass(frozen=True)
class Codec(Generic[T_co]):
    schema: Value
    accepts: Callable[[Value], bool]


def codec(schema: Value, accepts: Callable[[Value], bool]) -> Codec[T]:
    return Codec(copy_value(schema), accepts)


class Schemas:
    any: Codec[Value] = Codec({}, lambda value: True)
    string: Codec[str] = Codec({"type": "string"}, lambda value: type(value) is str)
    number: Codec[int | float] = Codec(
        {"type": "number"}, lambda value: type(value) in (int, float)
    )
    boolean: Codec[bool] = Codec({"type": "boolean"}, lambda value: type(value) is bool)

    @staticmethod
    def array(item: Codec[T]) -> Codec[list[T]]:
        return Codec(
            {"type": "array", "items": item.schema},
            lambda value: type(value) is list and all(item.accepts(v) for v in value),
        )

    @staticmethod
    def nullable(item: Codec[T]) -> Codec[T | None]:
        return Codec(
            {"anyOf": [item.schema, {"type": "null"}]},
            lambda value: value is None or item.accepts(value),
        )

    @staticmethod
    def object(fields: dict[str, Codec[Value]]) -> Codec[dict[str, Value]]:
        return Codec(
            {
                "type": "object",
                "properties": {k: c.schema for k, c in fields.items()},
                "required": list(fields),
                "additionalProperties": False,
            },
            lambda value: (
                type(value) is dict
                and set(value) == set(fields)
                and all(c.accepts(value[k]) for k, c in fields.items())
            ),
        )


s = Schemas()

File: packages/sdk/src/values.ts
export type Value = null | boolean | number | string | Value[] | { [key: string]: Value };
export type Code = import('./control.js').ErrorCode;
const actions: Record<Code, string> = {
  INVALID_PLAN: 'Check flow references, dependency limits and policy settings.',
  INVALID_VALUE: 'Use supported portable values and the declared operation schema.',
  UNSUPPORTED: 'Use an implemented capability and matching SDK/native versions.',
  CAPACITY: 'Reduce admitted work or explicitly adjust the runtime capacity.',
  OPERATION_FAILED:
    'Inspect the operation in the application; retry only under its declared semantics.',
  TRANSFORM_FAILED: 'Inspect the pure transformation at the indicated source.',
  CANCELLED: 'Start another invocation if the result is still needed.',
  TIMEOUT: 'Inspect queue and service time before changing the deadline.',
  CLOSED: 'Create a new Runtime for subsequent invocations.',
};
export class ClearingsError extends Error {
  readonly nextAction: string;
  constructor(
    readonly code: Code,
    readonly source: string | null = null,
  ) {
    super(`${code}${source ? ` at ${source}` : ''}. ${actions[code]}`);
    this.name = 'ClearingsError';
    this.nextAction = actions[code];
  }
}
export function nativeError(error: unknown): ClearingsError {
  if (error instanceof ClearingsError) return error;
  try {
    const parsed = JSON.parse(error instanceof Error ? error.message : String(error));
    if (Object.hasOwn(actions, parsed.code))
      return new ClearingsError(parsed.code, parsed.source ?? null);
  } catch {
    /* The native error may not be a serialized Failure. */
  }
  return new ClearingsError('INVALID_PLAN');
}
/** Snapshot the supported value domain without serializing application data through Rust. */
export function copyValue<T extends Value>(input: T): T {
  let entries = 0;
  let bytes = 0;
  const path = new Set<object>();
  function string(value: string): string {
    if (!value.isWellFormed()) throw new ClearingsError('INVALID_VALUE');
    bytes += Buffer.byteLength(value, 'utf8');
    if (bytes > 1_048_576) throw new ClearingsError('CAPACITY');
    return value;
  }
  function visit(value: unknown, depth: number): Value {
    if (++entries > 100_000 || depth > 64) throw new ClearingsError('CAPACITY');
    if (value === null || typeof value === 'boolean') return value;
    if (typeof value === 'string') return string(value);
    if (typeof value === 'number') {
      if (!Number.isFinite(value) || (Number.isInteger(value) && !Number.isSafeInteger(value)))
        throw new ClearingsError('INVALID_VALUE');
      return Object.is(value, -0) ? 0 : value;
    }
    if (typeof value !== 'object' || path.has(value)) throw new ClearingsError('INVALID_VALUE');
    path.add(value);
    let out: Value;
    if (Array.isArray(value)) {
      if (Object.getPrototypeOf(value) !== Array.prototype || value.length > 100_000)
        throw new ClearingsError('INVALID_VALUE');
      const values: Value[] = [];
      for (let i = 0; i < value.length; i++) {
        const descriptor = Object.getOwnPropertyDescriptor(value, String(i));
        if (!descriptor || !('value' in descriptor)) throw new ClearingsError('INVALID_VALUE');
        values.push(visit(descriptor.value, depth + 1));
      }
      out = values;
    } else {
      const prototype = Object.getPrototypeOf(value);
      if (
        (prototype !== Object.prototype && prototype !== null) ||
        Object.getOwnPropertySymbols(value).length
      )
        throw new ClearingsError('INVALID_VALUE');
      const result: Record<string, Value> = {};
      for (const [key, descriptor] of Object.entries(Object.getOwnPropertyDescriptors(value))) {
        if (!descriptor.enumerable || !('value' in descriptor))
          throw new ClearingsError('INVALID_VALUE');
        Object.defineProperty(result, string(key), {
          value: visit(descriptor.value, depth + 1),
          enumerable: true,
          writable: true,
          configurable: true,
        });
      }
      out = result;
    }
    path.delete(value);
    return out;
  }
  return visit(input, 0) as T;
}
export interface Codec<T extends Value> {
  readonly schema: Value;
  accepts(value: Value): value is T;
}
export function codec<T extends Value>(
  schema: Value,
  accepts: (value: Value) => value is T,
): Codec<T> {
  return { schema: copyValue(schema), accepts };
}
type Decoded<C> = C extends Codec<infer T> ? T : never;
export const s = {
  any: codec<Value>({}, (v): v is Value => true),
  string: codec<string>({ type: 'string' }, (v): v is string => typeof v === 'string'),
  number: codec<number>({ type: 'number' }, (v): v is number => typeof v === 'number'),
  boolean: codec<boolean>({ type: 'boolean' }, (v): v is boolean => typeof v === 'boolean'),
  array<T extends Value>(item: Codec<T>): Codec<T[]> {
    return codec(
      { type: 'array', items: item.schema },
      (v): v is T[] => Array.isArray(v) && v.every((x) => item.accepts(x)),
    );
  },
  nullable<T extends Value>(item: Codec<T>): Codec<T | null> {
    return codec(
      { anyOf: [item.schema, { type: 'null' }] },
      (v): v is T | null => v === null || item.accepts(v),
    );
  },
  object<C extends Record<string, Codec<Value>>>(
    fields: C,
  ): Codec<{ [K in keyof C]: Decoded<C[K]> }> {
    return codec(
      {
        type: 'object',
        properties: Object.fromEntries(Object.entries(fields).map(([k, c]) => [k, c.schema])),
        required: Object.keys(fields),
        additionalProperties: false,
      },
      (v): v is { [K in keyof C]: Decoded<C[K]> } =>
        v !== null &&
        typeof v === 'object' &&
        !Array.isArray(v) &&
        Object.keys(v).length === Object.keys(fields).length &&
        Object.entries(fields).every(([k, c]) => Object.hasOwn(v, k) && c.accepts(v[k]!)),
    );
  },
};
